cli/cli_app.go

```diff
@@ -542,26 +542,44 @@ func runTerragruntWithConfig(terragruntOptions *options.TerragruntOptions, terra
 		return err
 	}
 
-	return runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() error {
+	err, output := runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() (error, string) {
 		return runTerraformWithRetry(terragruntOptions)
 	})
+	if err != nil && isErrorRequiringInit(output, err, terragruntOptions) {
+		initOptions := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+		initOptions.TerraformCommand = "init"
+		initOptions.TerraformCliArgs = []string{"init"}
+		initOptions.AutoInit = false
+		if err := runTerragruntWithConfig(initOptions, terragruntConfig, allowSourceDownload); err != nil {
+			return err
+		}
+
+		err, _ = runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() (error, string) {
+			return runTerraformWithRetry(terragruntOptions)
+		})
```

Comment:
I'm a bit worried about this copy/pasted block. If we change something in the main one, we may forget to change something in this retry. Any way to structure the code to avoid the duplication?

Reply:
Let me see if I can make this better. Just to clarify - the "main one" is the first invocation of `runActionWithHooks` above this one that you are referring to?

Reply:
This is going to change.